src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -3450,83 +3467,112 @@ private SqlFedAuthToken getFedAuthToken(SqlFedAuthInfo fedAuthInfo) throws SQLSe
         // No:of attempts, for tracing purposes, if we underwent retries.
         int numberOfAttempts = 0;
 
-        FedAuthDllInfo dllInfo = null;
-
         String user = activeConnectionProperties.getProperty(SQLServerDriverStringProperty.USER.toString());
         String password = activeConnectionProperties.getProperty(SQLServerDriverStringProperty.PASSWORD.toString());
 
-        long expirationFileTime = 0;
-
         while (true) {
             numberOfAttempts++;
 
-            try {
-                if (authenticationString.trim().equalsIgnoreCase(SqlAuthentication.ActiveDirectoryPassword.toString())) {
-                    dllInfo = AuthenticationJNI.getAccessToken(user, password, fedAuthInfo.stsurl, fedAuthInfo.spn, clientConnectionId.toString(),
-                            ActiveDirectoryAuthentication.jdbcFedauthClientId, expirationFileTime);
+            if (authenticationString.trim().equalsIgnoreCase(SqlAuthentication.ActiveDirectoryPassword.toString())) {
+                ExecutorService executorService = Executors.newFixedThreadPool(1);
+                try {
+                    AuthenticationContext context = new AuthenticationContext(fedAuthInfo.stsurl, false, executorService);
+                    Future<AuthenticationResult> future = context.acquireToken(fedAuthInfo.spn, ActiveDirectoryAuthentication.jdbcFedauthClientId,
+                            user, password, null);
+
+                    AuthenticationResult authenticationResult = future.get();
+                    fedAuthToken = new SqlFedAuthToken(authenticationResult.getAccessToken(), authenticationResult.getExpiresOnDate());
+
+                    // Break out of the retry loop in successful case.
+                    break;
                 }
-                else if (authenticationString.trim().equalsIgnoreCase(SqlAuthentication.ActiveDirectoryIntegrated.toString())) {
-                    dllInfo = AuthenticationJNI.getAccessTokenForWindowsIntegrated(fedAuthInfo.stsurl, fedAuthInfo.spn, clientConnectionId.toString(),
-                            ActiveDirectoryAuthentication.jdbcFedauthClientId, expirationFileTime);
+                catch (MalformedURLException | InterruptedException e) {
+                    throw new SQLServerException(e.getMessage(), null);
                 }
+                catch (ExecutionException e) {
+                    MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ADALExecution"));
+                    Object[] msgArgs = {user, authenticationString};
 
-                // AccessToken should not be null.
-                assert null != dllInfo.accessTokenBytes;
+                    // the cause error message uses \\n\\r which does not give correct format
+                    // change it to \r\n to provide correct format
+                    String correctedErrorMessage = e.getCause().getMessage().replaceAll("\\\\r\\\\n", "\r\n");
```

Comment:
Do we have any Negative JUnit Test cases ? Which will expect some exception / message / cause.